library/constant_time_internal.h

```diff
@@ -20,224 +20,425 @@
 #ifndef MBEDTLS_CONSTANT_TIME_INTERNAL_H
 #define MBEDTLS_CONSTANT_TIME_INTERNAL_H
 
+#include <stdint.h>
+#include <stddef.h>
+
 #include "common.h"
 
 #if defined(MBEDTLS_BIGNUM_C)
 #include "mbedtls/bignum.h"
 #endif
 
-#if defined(MBEDTLS_SSL_TLS_C)
-#include "ssl_misc.h"
-#endif
+/* The constant-time interface provides various operations that are likely
+ * to result in constant-time code that does not branch or use conditional
+ * instructions for secret data (for secret pointers, this also applies to
+ * the data pointed to).
+ *
+ * It has three main parts:
+ *
+ * - boolean operations
+ *   These are all named mbedtls_ct_bool_<operation>, and operate over
+ *   mbedtls_ct_condition_t.
+ *   All arguments are considered secret.
+ *   example: bool x = y | z          =>    x = mbedtls_ct_bool_or(y, z)
```

Comment:
So yes, the comment isn't great, and yes, these functions all return bools (`mbedtls_ct_condition_t`).

But logically (see what I did there?) they are different: some manipulate boolean values, generally to combine them, and some process integers.

We could drop the `bool` part, but that would make extension of the API difficult in the future should it be necessary. If we wanted to do constant-time comparison of other integer types than `mbedtls_ct_uint_t`, having `uint` in the existing names would lead to an obvious naming strategy for future functions.

If @mpg prefers to just drop the `bool` part then fair enough, but I'd rather see `bool` and `uint` in the names.

Reply:
I tend to agree with Tom here. IMO when presenting operations in a language, it's customary (I'll give [this wikipedia page](https://en.wikipedia.org/wiki/Operators_in_C_and_C%2B%2B) as an example but I think we could find several more) to distinguish comparison operators from logical operators.

Also, I think functions are more often named after the type of their inputs than outputs - when types are included in the function name at all. Here I think it depends whether we think we're likely to have `eq` and `lt` operators for other types that `uint`, which does not seem likely to me (it's customary in C for the comparison operator for byte arrays to be called `memcmp` - and that's the case in our constant-time API too).

So, I have a slight preference for `ct_eq` and `ct_lt` over `ct_uint_eq` and `ct_uint_lt`, and a rather stronger preference for any of these names over the current `ct_bool_eq` and `ct_bool_lt`.

Just to be clear, I'm happy with `ct_bool_not`, `ct_bool_or`, `ct_bool_and` and I think we should keep those names.

Reply:
PS: while talking names, I think I already wrote that I'd prefer `if_else0` to `if0`, but we can also just go with the current name, and see if we get used to it or it keeps bugging us - and then renaming a purely internal function is a very easy thing to do.

Reply:
@mpg Surely If we're going to keep `bool` in `ct_bool_not` etc, it seems better to have `uint` in the `_eq` and other comparison functions so that the names are of the form `ct` `_` TYPE `_` OPERATTION, rather than some having TYPE `bool` and some not having a type in there

Reply:
On the basis that @mpg is happy with the current names, I'm not going to iterate on this today. I'm happy to re-visit with a name-change PR in the future, if we find that it bothers us.

@tom-cosgrove-arm if you have no other comments, let's approve and merge please

Reply:
Note that it would be perfectly reasonable to have a function to check bools for equality - it's NOT XOR. So `ct_bool_eq` SHOULD be !`ct_bool_xor`, and we SHOULD have `ct_uint_eq`. Which means the other `mbedtls_ct_uint_t` comparison functions should have `_uint_` in their names